sound/soc/sof/intel/hda.c

```diff
@@ -543,7 +543,8 @@ int hda_dsp_probe(struct snd_sof_dev *sdev)
 	 * if it fails, use legacy interrupt mode
 	 * TODO: support msi multiple vectors
 	 */
-	if (hda_use_msi && !pci_alloc_irq_vectors(pci, 1, 1, PCI_IRQ_MSI)) {
+	ret = pci_alloc_irq_vectors(pci, 1, 1, PCI_IRQ_MSI);
+	if (hda_use_msi && ret > 0) {
```

Comment:
if (hda_use_msi && pci_alloc_irq_vectors(pci, 1, 1, PCI_IRQ_MSI) > 0) {

Reply:
my question is: should irq_vectors be released if it is not used ? or do it like keyon's suggestion ?

Reply:
> if (hda_use_msi && pci_alloc_irq_vectors(pci, 1, 1, PCI_IRQ_MSI) > 0) {

testing the return value inside of the if() is the reason why we got into this mess in the first place. I prefer to have an intermediate variable that can be traced for debug easily.

Reply:
> my question is: should irq_vectors be released if it is not used ? or do it like keyon's suggestion ?

They are already released, see multiple calls to 

````c
free_irq_vector:
	if (sdev->msi_enabled)
		pci_free_irq_vectors(pci);

hda_dsp_remove()
{
...
	free_irq(sdev->ipc_irq, sdev);
	free_irq(hda->irq, bus);
	if (sdev->msi_enabled)
		pci_free_irq_vectors(pci);
...
}

Reply:
if hda_use_msi == false, sdev->msi_enable = false, so pci_free_irq_vectors(pci) will not be invoked.

Reply:
> if hda_use_msi == false, sdev->msi_enable = false, so pci_free_irq_vectors(pci) will not be invoked.

@RanderWang I am sorry, but why do we need pci_free_irq_vectors(pci) if sdev->msi_enable is false?